runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPipelineFuser.java

```diff
@@ -187,11 +187,19 @@ private DescendantConsumers getRootConsumers(PTransformNode rootNode) {
         rootNode.getId(),
         rootNode.getTransform().getInputsMap());
     checkArgument(
-        !pipeline.getEnvironment(rootNode).isPresent(),
-        "%s requires all root nodes to be runner-implemented %s primitives, "
+        !pipeline.getEnvironment(rootNode).isPresent()
+            // We allow Read transforms as root transforms. The Runner can choose whether it
```

Comment:
I am not super familiar with this piece of code so let me know if this is trivial.
Will this bypass the cross language read as this will execute the transform on Flink natively and not on Java SDKHarness?

Reply:
Root transforms are enforced to be environment-free and input-free transforms, i.e. Impulse. Read executes in an environment and is executed by the harness, but in order to support bounded and unbounded reads, we need to translate it natively. Otherwise we are limited to bounded sources.

Reply:
So if a Python Read leaked through, this would try and execute it? It does feel a bit odd to augment the fuser to know about Read--better to augment the fuser to accept a list of known native transforms from the runner invoking it.